expat/lib/xmlparse.c

```diff
@@ -3018,6 +3022,8 @@ doContent(XML_Parser parser, int startTagLevel, const ENCODING *enc,
         len = XmlNameLength(enc, rawName);
         if (len != tag->rawNameLength
             || memcmp(tag->rawName, rawName, len) != 0) {
+          moveToFreeBindingList(parser, tag->bindings);
+          tag->bindings = NULL;
```

Comment:
So far, I cannot confirm of a leak. This line of code is touch by the test suite and test suite is executed under LeakSan. I would expect a call to `XML_ParserFree` to "plug" this leak, and missing to call `XML_ParserFree` would be a bug in the code using Expat. Please help me understand why we have a true leak here if so, thank you.

Reply:
Here we move tag from m_tagStack to m_freeTagList:

3017:
```
  TAG *tag = parser->m_tagStack;
  parser->m_tagStack = tag->parent;
  tag->parent = parser->m_freeTagList;
  parser->m_freeTagList = tag;
```

We don't do anything with the tag bindings until later:

3057:
```
  while (tag->bindings) {
    BINDING *b = tag->bindings;
    if (parser->m_endNamespaceDeclHandler)
      parser->m_endNamespaceDeclHandler(parser->m_handlerArg,
                                        b->prefix->name);
    tag->bindings = tag->bindings->nextTagBinding;
    b->nextTagBinding = parser->m_freeBindingList;
    parser->m_freeBindingList = b;
    b->prefix->binding = b->prevPrefixBinding;
  }
```

This means that we can have a TAG object on  m_freeTagList with a non-NULL
bindings pointer.

If the user then calls XML_ParserFree, this will be handled correctly:

1439:
```
  /* free m_tagStack and m_freeTagList */
  tagList = parser->m_tagStack;
  for (;;) {
    TAG *p;
    if (tagList == NULL) {
      if (parser->m_freeTagList == NULL)
        break;
      tagList = parser->m_freeTagList;
      parser->m_freeTagList = NULL;
    }
    p = tagList;
    tagList = tagList->parent;
    FREE(parser, p->buf);
    destroyBindings(p->bindings, parser);
    FREE(parser, p);
  }
```

However, if the user instead calls XML_ParserReset, which the API documentation
suggests to me is an intended use-case, then this is not handled:

1186:
```
  XML_Bool XMLCALL
  XML_ParserReset(XML_Parser parser, const XML_Char *encodingName) {
    TAG *tStk;
    OPEN_INTERNAL_ENTITY *openEntityList;

    if (parser == NULL)
      return XML_FALSE;

    if (parser->m_parentParser)
      return XML_FALSE;
    /* move m_tagStack to m_freeTagList */
    tStk = parser->m_tagStack;
    while (tStk) {
      TAG *tag = tStk;
      tStk = tStk->parent;
      tag->parent = parser->m_freeTagList;
      moveToFreeBindingList(parser, tag->bindings);
      tag->bindings = NULL;
      parser->m_freeTagList = tag;
    }
    /* move m_openInternalEntities to m_freeInternalEntities */
    openEntityList = parser->m_openInternalEntities;
    while (openEntityList) {
      OPEN_INTERNAL_ENTITY *openEntity = openEntityList;
      openEntityList = openEntity->next;
      openEntity->next = parser->m_freeInternalEntities;
      parser->m_freeInternalEntities = openEntity;
    }
    moveToFreeBindingList(parser, parser->m_inheritedBindings);
    FREE(parser, parser->m_unknownEncodingMem);
    if (parser->m_unknownEncodingRelease)
      parser->m_unknownEncodingRelease(parser->m_unknownEncodingData);
    poolClear(&parser->m_tempPool);
    poolClear(&parser->m_temp2Pool);
    FREE(parser, (void *)parser->m_protocolEncodingName);
    parser->m_protocolEncodingName = NULL;
    parserInit(parser, encodingName);
    dtdReset(parser->m_dtd, &parser->m_mem);
    return XML_TRUE;
  }
```

When we then later re-use the tag, the bindings pointer is just cleared:

2904:
```
  TAG *tag;
  enum XML_Error result;
  XML_Char *toPtr;
  if (parser->m_freeTagList) {
    tag = parser->m_freeTagList;
    parser->m_freeTagList = parser->m_freeTagList->parent;
  } else {
    tag = (TAG *)MALLOC(parser, sizeof(TAG));
    if (! tag)
      return XML_ERROR_NO_MEMORY;
    tag->buf = (char *)MALLOC(parser, INIT_TAG_BUF_SIZE);
    if (! tag->buf) {
      FREE(parser, tag);
      return XML_ERROR_NO_MEMORY;
    }
    tag->bufEnd = tag->buf + INIT_TAG_BUF_SIZE;
  }
  tag->bindings = NULL;
```

AFAIU this won't happen without a call to XML_ParserReset, since we
need to go through the error path in XML_TOK_END_TAG, and at that point
we'll be in an error state and shouldn't continue using the parser.

(Another fix would be to have XML_ParserReset add the bindings from all
of the free tags to m_freeBindingsList, but given that the path which 
allocates from m_freeTagList clearly doesn't expect this pointer to be
valid, I thought it was better to ensure that invariant is always true
instead.)

Reply:
@c01db33f excellent explanation, thank you! :+1: I confirm that this is a true memory leak bug.

I believe I have found an even simpler fix: We could postpone the move to the free tag list until after the closing tag name match check. I have created a pull request #653 now that contains that approach of a patch and a reproducer test based on your explanation. You review would be appreciated.

(I have also created a dedicated ticket for this issue at #652.)

Reply:
Related: https://github.com/libexpat/libexpat/pull/653#issuecomment-1271789796